web/packages/teleport/src/Notifications/Notifications.tsx

```diff
@@ -0,0 +1,343 @@
+/**
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState, useEffect } from 'react';
+import { isBefore } from 'date-fns';
+import styled from 'styled-components';
+import { Alert, Box, Flex, Indicator, Text } from 'design';
+
+import { Notification as NotificationIcon, BellRinging } from 'design/Icon';
+import { useRefClickOutside } from 'shared/hooks/useRefClickOutside';
+import useAttempt from 'shared/hooks/useAttemptNext';
+import { HoverTooltip } from 'shared/components/ToolTip';
+
+import { useTeleport } from 'teleport';
+import useStickyClusterId from 'teleport/useStickyClusterId';
+import { Dropdown } from 'teleport/components/Dropdown';
+
+import { ButtonIconContainer } from 'teleport/TopBar/Shared';
+import {
+  FetchNotificationsResponse,
+  Notification as NotificationType,
+} from 'teleport/services/notifications';
+import { Notification } from './Notification';
+import { useInfiniteScroll } from 'shared/hooks';
+
+const PAGE_SIZE = 15;
+
+export function Notifications({ iconSize = 24 }: { iconSize?: number }) {
+  const ctx = useTeleport();
+  const { clusterId } = useStickyClusterId();
+
+  const { attempt, run } = useAttempt();
+
+  const [response, setResponse] = useState<
+    Omit<FetchNotificationsResponse, 'notifications'>
+  >({
+    userNotificationsNextKey: '',
+    globalNotificationsNextKey: '',
+    userLastSeenNotification: undefined,
+  });
+  const [notifications, setNotifications] = useState<NotificationType[]>([]);
+
+  function fetchNotifications(): Promise<any> {
+    return run(() =>
+      ctx.notificationService
+        .fetchNotifications({
+          clusterId,
+          userNotificationsStartKey: response.userNotificationsNextKey,
+          globalNotificationsStartKey: response.globalNotificationsNextKey,
+          limit: PAGE_SIZE,
+        })
+        .then(({ notifications: newNotifications, ...resp }) => {
+          setNotifications([...notifications, ...newNotifications]);
```

Comment:
It would be great if we could reuse `useKeyBasedPagination` here. Fetching data in infinite scroll isn't that simple task and Bartosz spent some time to make it work reliably (I mean handling concurrent requests, aborting previous requests, etc).
I'm a little afraid that this implementation may be prone to race conditions as it stores its state in three variables.
I also noticed that sometimes the component makes many identical requests to get the notifications (because it doesn't handle concurrent requests).

Here is my attempt to switch to `useKeyBasedPagination`, let me know what you think about it.
Because the data doesn't fully match the shape that the hook expects, I had to make some workarounds.
```
Subject: [PATCH] Use `useKeyBasedPagination`
---
Index: web/packages/teleport/src/Notifications/Notifications.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/teleport/src/Notifications/Notifications.tsx b/web/packages/teleport/src/Notifications/Notifications.tsx
--- a/web/packages/teleport/src/Notifications/Notifications.tsx	(revision aed5707540012e8c1aee26975e7a2ef4e211ccfa)
+++ b/web/packages/teleport/src/Notifications/Notifications.tsx	(date 1713871439685)
@@ -16,85 +16,79 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import React, { useState, useEffect } from 'react';
+import React, { useState, useCallback } from 'react';
 import { isBefore } from 'date-fns';
 import styled from 'styled-components';
 import { Alert, Box, Flex, Indicator, Text } from 'design';
 
 import { Notification as NotificationIcon, BellRinging } from 'design/Icon';
 import { useRefClickOutside } from 'shared/hooks/useRefClickOutside';
-import useAttempt from 'shared/hooks/useAttemptNext';
 import { HoverTooltip } from 'shared/components/ToolTip';
 
+import { useInfiniteScroll } from 'shared/hooks';
+
+import { useKeyBasedPagination } from 'shared/hooks/useInfiniteScroll';
+
 import { useTeleport } from 'teleport';
 import useStickyClusterId from 'teleport/useStickyClusterId';
 import { Dropdown } from 'teleport/components/Dropdown';
 
 import { ButtonIconContainer } from 'teleport/TopBar/Shared';
-import {
-  FetchNotificationsResponse,
-  Notification as NotificationType,
-} from 'teleport/services/notifications';
+
 import { Notification } from './Notification';
-import { useInfiniteScroll } from 'shared/hooks';
 
 const PAGE_SIZE = 15;
+const START_KEY_SEPARATOR = ' ';
 
 export function Notifications({ iconSize = 24 }: { iconSize?: number }) {
   const ctx = useTeleport();
   const { clusterId } = useStickyClusterId();
 
-  const { attempt, run } = useAttempt();
-
-  const [response, setResponse] = useState<
-    Omit<FetchNotificationsResponse, 'notifications'>
-  >({
-    userNotificationsNextKey: '',
-    globalNotificationsNextKey: '',
-    userLastSeenNotification: undefined,
-  });
-  const [notifications, setNotifications] = useState<NotificationType[]>([]);
-
-  function fetchNotifications(): Promise<any> {
-    return run(() =>
-      ctx.notificationService
-        .fetchNotifications({
+  const { resources, fetch, attempt } = useKeyBasedPagination({
+    fetchMoreSize: PAGE_SIZE,
+    initialFetchSize: PAGE_SIZE,
+    fetchFunc: useCallback(
+      async paginationParams => {
+        // Separate the keys.
+        const startKeys = paginationParams.startKey.split(START_KEY_SEPARATOR);
+        const response = await ctx.notificationService.fetchNotifications({
           clusterId,
-          userNotificationsStartKey: response.userNotificationsNextKey,
-          globalNotificationsStartKey: response.globalNotificationsNextKey,
-          limit: PAGE_SIZE,
-        })
-        .then(({ notifications: newNotifications, ...resp }) => {
-          setNotifications([...notifications, ...newNotifications]);
-          setResponse(resp);
+          userNotificationsStartKey: startKeys[0],
+          globalNotificationsStartKey: startKeys[1],
+          limit: paginationParams.limit,
+        });
 
-          // If we've reached the end of both lists, hide the infinite scroll trigger.
-          if (
-            resp.globalNotificationsNextKey === '' &&
-            resp.userNotificationsNextKey === ''
-          ) {
-            setRenderTrigger(false);
-          }
-        })
-    );
-  }
+        // We can't pass notifications directly because we need the full
+        // response object to inspect `userLastSeenNotification`.
+        return {
+          agents: response.notifications.length ? [response] : [],
+          startKey:
+            response.userNotificationsNextKey ||
+            response.globalNotificationsNextKey
+              ? `${response.userNotificationsNextKey}${START_KEY_SEPARATOR}${response.globalNotificationsNextKey}`
+              : undefined,
+        };
+      },
+      [clusterId, ctx.notificationService]
+    ),
+  });
 
-  useEffect(() => {
-    fetchNotifications();
-  }, []);
+  // flatten the items
+  const notifications = resources.flatMap(r => r.notifications);
+  // take the most recent item
+  const userLastSeenNotification = resources.at(-1)?.userLastSeenNotification;
 
   const { setTrigger } = useInfiniteScroll({
-    fetch: fetchNotifications,
+    fetch,
   });
 
   const [view, setView] = useState<View>('All');
   const [open, setOpen] = useState(false);
-  const [renderTrigger, setRenderTrigger] = useState(true);
 
   const ref = useRefClickOutside<HTMLDivElement>({ open, setOpen });
 
   const unseenNotifsCount = notifications.filter(notif =>
-    isBefore(response.userLastSeenNotification, notif.createdDate)
+    isBefore(userLastSeenNotification, notif.createdDate)
   ).length;
 
   const items = notifications
@@ -143,12 +137,8 @@
         {attempt.status === 'success' && items.length === 0 && <EmptyState />}
         <NotificationsList>
           <>
-            {!!items.length && (
-              <>
-                {items}
-                {open && renderTrigger && <div ref={setTrigger} />}
-              </>
-            )}
+            {!!items.length && items}
+            {open && <div ref={setTrigger} />}
             {attempt.status === 'processing' && (
               <Flex
                 width="100%"
@@ -279,6 +269,7 @@
       : props.theme.colors.text.muted};
   background-color: ${props =>
     props.selected ? props.theme.colors.brand : 'transparent'};
+
   .selected {
     color: ${props => props.theme.colors.text.primaryInverse};
     background-color: ${props => props.theme.colors.brand};
```